### PR TITLE
# 155769736 Restrict API Access by Roles

### PR DESCRIPTION
**Pull request by Andretalik**
## Description (what problem you're fixing)

  - At the moment, all endpoints can be used by any Andelan. This should not be the case seeing as to how there are certain things that only specific people are authorised to do, i.e: Only SuccessOps have the jurisdiction and authority to approve points submission by Society secretaries.

## Fix (what you did to fix it)

  - Created a decorator that checks the token supplied to the logged in user for the roles of the user and grant access to endpoints based off of their roles.
 
## How to test (describe how to test your PR)

  - Run `python manage.py test` on a clone of the repository
  - Using Postman, put in all the required headers and attempt to use an endpoint in which that type if user is not authorised.

### CODE REVIEW

**Comment by georgreen** [comment]
Rebase your work.

File: api/initial_data.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+# roles available
+roles = [Role(uuid="-KXGy1EB1oimjQgFim6F", name="Success"),
+         Role(uuid="-KXGy1EB1oimjQgFim6L", name="Finance"),
+         Role(uuid="-KXGy1EB1oimjQgFim6C", name="Fellow"),
+         Role(uuid="-KkLwgbeJUO0dQKsEk1i", name="Success Ops"),
+         Role(uuid="-KiihfZoseQeqC6bWTau", name="Andelan")]
+
+# LoggingActivity
+blog_points = LoggedActivity(value=blog.value,
+<<<<<<< HEAD
```
Comment (LarryWachira) [inline api/initial_data.py]: Seems like you missed some diff markers here...

File: api/models.py [hunk 2]
```diff
@@ -110,19 +115,42 @@ class Cohort(Base):
...
     __tablename__ = 'users'
-    name= db.Column(db.String, nullable=False)
+    name = db.Column(db.String, nullable=False)
     email = db.Column(db.String, nullable=False, unique=True)
     role = db.Column(db.String, default="member")
 
     society_id = db.Column(db.String, db.ForeignKey('societies.uuid'))
     country_id = db.Column(db.String, db.ForeignKey('countries.uuid'))
+<<<<<<< HEAD
+<<<<<<< HEAD
```
Comment (LarryWachira) [inline api/models.py]: A looot of markers....

File: api/models.py [hunk 3]
```diff
@@ -110,19 +115,42 @@ class Cohort(Base):
...
 
 
 class User(Base):
     """Models Users."""
 
     __tablename__ = 'users'
-    name= db.Column(db.String, nullable=False)
+    name = db.Column(db.String, nullable=False)
     email = db.Column(db.String, nullable=False, unique=True)
     role = db.Column(db.String, default="member")
```
Comment (LarryWachira) [inline api/models.py]: This should be a foreign key now right?

File: tests/test_role.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+"""Test suite for restriction of endpoints by Role."""
+
+from .base_test import BaseTestCase
+
+
+class RoleTestCase(BaseTestCase):
+    """Holds tests of roles in authentication."""
+
+    def test_andelan_can_access_(self):
+        pass
```
Comment (LarryWachira) [inline tests/test_role.py]: I imagine this is incomplete.

**Comment by LarryWachira** [comment]
You also need to rebase... 59 commits is a lot.

File: tests/test_role.py (hunk 4 above)
Comment (Andretalik) [reply tests/test_role.py]: @LarryWachira Still quite a lot of confusion due to the attempted rebase of the work.
All these shall be resolved ASAP

File: tests/test_activities.py [hunk 5]
```diff
@@ -1,10 +1,85 @@
+"""Test of Activities Module."""
 import json
```
Comment (georgreen) [inline tests/test_activities.py]: Did you introduce this changes? I thought this belonged to @jonathankamau .

File: .gitignore [hunk 6]
```diff
@@ -103,6 +108,9 @@ ENV/
 .mypy_cache/
 
 .vscode/
-test_db.sqlite
+*.sqlite
+
+# Databases(SQLite)
+*.sqlite
```
Comment (jonathankamau) [inline .gitignore]: You have `*.sqlite` twice on line 111 and 114. One of them can be removed.

File: tests/test_role.py (hunk 4 above)
Comment (Andretalik) [reply tests/test_role.py]: It has bee completed

File: api/models.py [hunk 7]
```diff
@@ -117,7 +116,8 @@ class User(Base):
     """Models Users."""
 
     __tablename__ = 'users'
-    email = db.Column(db.String)
+    name= db.Column(db.String, nullable=False)
+    email = db.Column(db.String, nullable=False, unique=True)
     role = db.Column(db.String, default="member")
```
Comment (LarryWachira) [inline api/models.py]: I believe we discussed that this should be a relationship.

File: api/utils/initial_data.py [hunk 8]
```diff
@@ -3,8 +3,12 @@
 
 This contains the sample initial data required for the test run of the system.
 """
-from .models import (ActivityType, Activity, Country, LoggedActivity, Society,
+from api.models import (ActivityType, Activity, Country, LoggedActivity, Society,
+<<<<<<< HEAD:api/utils/initial_data.py
```
[the remark this answers is not in the document]
Comment (LarryWachira) [reply api/utils/initial_data.py]: Missed a spot.

File: api/utils/auth.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            # now return wrapped function
+            return f(*args, **kwargs)
+    return decorated
+
+def roles_required(roles):  # roles should be a list
+    """Ensure only authorised roles may access sensitive data."""
+    def check_user_role(f):
+        @wraps(f)
+        def decorated(*args, **kwargs):
+            if g.current_user.society_position not in roles:
```
Comment (LarryWachira) [inline api/utils/auth.py]: We haven't used `society_position` anywhere else in the code. I doubt this will authorize well.

Comment (Andretalik) [reply api/utils/auth.py]: Rebasing is a headache

**Comment by Andretalik** [comment]
I am closing this PR and opening a new cleaner one

**Comment by georgreen** [comment]
@Andretalik What's the progress on this?

File: tests/test_activities.py (hunk 5 above)
Comment (Andretalik) [reply tests/test_activities.py]: Which will be reflected once his work is merged first.

**Comment by LarryWachira** [comment]
Can't review until you rebase.

File: src/api/utils/initial_data.py [hunk 10]
```diff
@@ -3,7 +3,7 @@
 This contains the sample initial data required for the test run of the system.
 """
 from api.models import (ActivityType, Activity, Country, LoggedActivity, Society,
-                        User)
+                        User, Role)
```
Comment (LarryWachira) [inline src/api/utils/initial_data.py]: No Cohort? Definitely missed something. Consider using the initial_data on `ft-log-activities`.

**Comment by georgreen** [comment]
@Andretalik I need to merge this Asap, can we resolve the conflicts?

File: src/tests/base_test.py [hunk 11]
```diff
@@ -123,7 +166,7 @@ def setUp(self):
             email="test.user.societies@andela.com",
             country=self.nigeria,
             cohort=self.cohort_1_Nig,
-            society = self.phoenix)
+            society=self.phoenix)
 
         self.test_user_2 = User(
             uuid="-KdQsawesome_useridZ",
```
Comment (LarryWachira) [inline src/tests/base_test.py]: Get rid of `role="president"` at line 187 to fix the error. Also use the newer `initial_data.py` you shared on DM and make the minor changes.

File: src/tests/test_activities.py [hunk 12]
```diff
@@ -7,6 +7,14 @@
...
+        """Save required roles."""
+        BaseTestCase.setUp(self)
+        self.successops_role.save()
+        self.fellow_role.save()
+        self.success_role.save()
+        self.finance_role.save()
+
     def test_create_activity(self):
         """Test that an activity has been created successfully."""
         new_activity = dict(name="tech congress",
```
Comment (LarryWachira) [inline src/tests/test_activities.py]: Could you also change the hard-coded `activityDate`s in this file? Jonathan missed them and these tests will be useless in 2020.

**Comment by Andretalik** [comment]
This PR is old and outdated hence the multiple breakages in logic and functionality.
This shall be closed and the work done placed on a cleaner branch.

**Pull request closed without merging**